core/src/main/java/org/apache/calcite/rel/rules/SortProjectPullUpConstantsRule.java

```diff
@@ -0,0 +1,176 @@
+        .stream()
+        .map(RelFieldCollation::getFieldIndex)
+        .collect(Collectors.toList());
+    if (fieldCollationIndexes.size() <= 0) {
+      return;
+    }
+
+    final RexBuilder rexBuilder = sort.getCluster().getRexBuilder();
+    final RelMetadataQuery mq = call.getMetadataQuery();
+    final RelOptPredicateList predicates =
+        mq.getPulledUpPredicates(input);
+    if (RelOptPredicateList.isEmpty(predicates)) {
+      return;
+    }
+
+    final NavigableMap<Integer, RexNode> map = new TreeMap<>();
+    for (int fieldCollationIndex : fieldCollationIndexes) {
+      final RexInputRef ref =
+          rexBuilder.makeInputRef(input, fieldCollationIndex);
+      if (predicates.constantMap.containsKey(ref)) {
+        map.put(fieldCollationIndex, predicates.constantMap.get(ref));
+      }
+    }
+
+    // None of the sort field expressions are constant. Nothing to do.
+    if (map.isEmpty()) {
+      return;
+    }
+
+    // Input of new top constant project.
+    final RelNode newInputRel;
+    if (map.size() == fieldCollationIndexes.size()) {
+      // All field collations are constants.
+      if (sort.offset == null && sort.fetch == null) {
+        // No any offset or fetch in current sort.
+        // Remove current sort.
+        newInputRel = input;
+      } else {
+        // Some offset or fetch exist in current sort.
+        newInputRel = sort.copy(sort.getTraitSet(), input, RelCollations.EMPTY);
+      }
+    } else {
+      // All field collations are not all constants.
+      final List<RelFieldCollation> newFieldCollations = new ArrayList<>();
+      for (RelFieldCollation fieldCollation : collation.getFieldCollations()) {
+        final int fieldIndex = fieldCollation.getFieldIndex();
+        if (map.containsKey(fieldIndex)) {
+          continue;
+        }
+        newFieldCollations.add(fieldCollation);
+      }
+      final RelCollation newRelCollation = RelCollations.of(newFieldCollations);
+      newInputRel = sort.copy(sort.getTraitSet(), input, newRelCollation);
+    }
+
+    final RelBuilder relBuilder = call.builder();
+    relBuilder.push(newInputRel);
+
+    // Create a projection back again.
+    List<Pair<RexNode, String>> projects = new ArrayList<>();
+    for (RelDataTypeField field : sort.getRowType().getFieldList()) {
+      final int index = field.getIndex();
+      final RexNode constantNode = map.get(index);
+      final RexNode newExpr;
+      if (constantNode == null) {
+        newExpr = rexBuilder.makeInputRef(newInputRel, index);
+      } else {
+        // Re-generate the constant expression in the project.
+        RelDataType originalType =
+            sort.getRowType().getFieldList().get(projects.size()).getType();
+        if (!originalType.equals(constantNode.getType())) {
+          newExpr = rexBuilder.makeCast(originalType, constantNode, true);
+        } else {
+          newExpr = constantNode;
+        }
```

Comment:
Having in the `if` clause a negated statement increases the cognitive complexity of the code, we can swap the `then/else` branches and remove the negation:

```suggestion
        if (originalType.equals(constantNode.getType())) {
          newExpr = constantNode;
        } else {
          newExpr = rexBuilder.makeCast(originalType, constantNode, true);
        }
```